httpserver/bin/note_server.dart

```diff
@@ -6,67 +6,85 @@
 // Use note_taker.html to run the client.
 
 import 'dart:io';
-import 'dart:convert' show JSON;
+import 'dart:async';
+import 'dart:convert' show UTF8, JSON;
 
 int count = 0;
 
-void main() {
+main() async {
   // One note per line.
-  List<String> lines = new File('notes.txt').readAsLinesSync();
-  count = lines.length;
+  try {
+    List<String> lines = new File('notes.txt').readAsLinesSync();
+    count = lines.length;
 
-  HttpServer.bind(InternetAddress.LOOPBACK_IP_V4, 4042)
-      .then(listenForRequests)
-      .catchError((e) => print('hello: ${e.toString()}'));
+    var server =
+        await HttpServer.bind(InternetAddress.LOOPBACK_IP_V4, 4042);
+    print('Listening for requests on 4042.');
+    await listenForRequests(server);
+  } on FileSystemException {
+    print('Could not open notes.txt');
+  } catch (e) {
+    print('hello: ${e.toString()}');
+  }
 }
 
-listenForRequests(_server) {
-  _server.listen((HttpRequest request) {
-    switch (request.method) {
-      case 'POST':
-        handlePost(request);
-        break;
-      case 'OPTION':
-        handleOptions(request);
-        break;
-      default:
-        defaultHandler(request);
-        break;
+Future listenForRequests(HttpServer requests) async {
+  try {
+    await for (HttpRequest request in requests) {
+      switch (request.method) {
+        case 'POST':
+          handlePost(request);
+          break;
+        case 'OPTION':
+          handleOptions(request);
+          break;
+        default:
+          defaultHandler(request);
+          break;
+      }
     }
-  },
-  onDone: () => print('No more requests.'),
-  onError: (e ) => print(e.toString()));
+    print('No more requests.');
+  } catch (e) {
+    print(e.toString());
+  }
 }
 
-void handlePost(HttpRequest req) {
-  StringBuffer data = new StringBuffer();
-
-  addCorsHeaders(req.response);
+handlePost(HttpRequest request) async {
+  addCorsHeaders(request.response);
 
-  req.listen((buffer) {
-    data.write(new String.fromCharCodes(buffer));
-  }, onDone: () {
-    var decoded = JSON.decode(data.toString());
+  try {
+    var decoded = await request
+        .transform(UTF8.decoder.fuse(JSON.decoder)).first;
 
     if (decoded.containsKey('myNote')) {
-      saveNote(req, '${decoded["myNote"]}\n');
-    } else { // 'getNote'
-      getNote(req, decoded['getNote']);
+      saveNote(request, '${decoded["myNote"]}\n');
+    } else {
+      getNote(request, decoded['getNote']);
     }
-  }, onError: (_) {
-    print('Request listen error.');
-  });
+  } catch (e) {
```

Comment:
This must come from the UTF8 or JSON decoders since saveNote and getNote both handle errors (or possibly from the request.respons, if that's possible).
Maybe move the catch to just after the transform?